core/src/main/java/com/linecorp/armeria/internal/common/ThreadLocalRequestContextStorage.java

```diff
@@ -16,67 +16,46 @@
 
 package com.linecorp.armeria.internal.common;
 
+import static com.linecorp.armeria.internal.common.RequestContextUtil.newIllegalContextPoppingException;
 import static java.util.Objects.requireNonNull;
 
 import javax.annotation.Nullable;
 
 import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestContextStorage;
 
 import io.netty.util.concurrent.FastThreadLocal;
 import io.netty.util.internal.InternalThreadLocalMap;
 
-public final class RequestContextThreadLocal {
+final class ThreadLocalRequestContextStorage implements RequestContextStorage {
 
     private static final FastThreadLocal<RequestContext> context = new FastThreadLocal<>();
 
-    /**
-     * Returns the current {@link RequestContext} in the thread-local.
-     */
     @Nullable
+    @Override
     @SuppressWarnings("unchecked")
-    public static <T extends RequestContext> T get() {
-        return (T) context.get();
-    }
-
-    /**
-     * Sets the specified {@link RequestContext} in the thread-local and returns the old {@link RequestContext}.
-     */
-    @Nullable
-    @SuppressWarnings("unchecked")
-    public static <T extends RequestContext> T getAndSet(RequestContext ctx) {
-        requireNonNull(ctx, "ctx");
-        final InternalThreadLocalMap map = InternalThreadLocalMap.get();
-        final RequestContext oldCtx = context.get(map);
-        context.set(map, ctx);
-        return (T) oldCtx;
-    }
-
-    /**
-     * Removes the {@link RequestContext} in the thread-local and returns it.
-     */
-    @Nullable
-    @SuppressWarnings("unchecked")
-    public static <T extends RequestContext> T getAndRemove() {
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
         final InternalThreadLocalMap map = InternalThreadLocalMap.get();
         final RequestContext oldCtx = context.get(map);
-        context.remove();
+        context.set(map, toPush);
         return (T) oldCtx;
     }
 
-    /**
-     * Sets the specified {@link RequestContext} in the thread-local.
-     */
-    public static void set(RequestContext ctx) {
-        requireNonNull(ctx, "ctx");
-        context.set(ctx);
+    @Override
+    public void pop(RequestContext current, @Nullable RequestContext toRestore) {
+        requireNonNull(current, "current");
+        final RequestContext contextInThreadLocal = context.get();
```

Comment:
Think you can optimize a but getting the map for both get / set like above.

Reply:
Done!